src/modules/wolfssl/nm_wolfssl_cli.c

```diff
@@ -0,0 +1,727 @@
+    }
+    return nm_wolfssl_util_fp_from_crt(crt, fp);
+}
+
+np_error_code set_handshake_timeout(struct np_dtls_cli_context* ctx, uint32_t minTimeout, uint32_t maxTimeout)
+{
+    int min = minTimeout / 1000;
+    int max = maxTimeout / 1000;
+    if (wolfSSL_dtls_set_timeout_init(ctx->ssl, min) !=
+        WOLFSSL_SUCCESS) {
+        NABTO_LOG_ERROR(LOG, "Cannot set min timeout for DTLS client connection");
+        return NABTO_EC_FAILED;
+    };
+    if (wolfSSL_dtls_set_timeout_max(ctx->ssl, max) != WOLFSSL_SUCCESS) {
+        NABTO_LOG_ERROR(LOG, "Cannot set max timeout for DTLS client connection");
+        return NABTO_EC_FAILED;
+    }
+    return NABTO_EC_OK;
+}
+
+/*
+ * asyncroniously start a dtls connection
+ */
+np_error_code nm_dtls_connect(struct np_dtls_cli_context* ctx)
+{
+    ctx->state = CONNECTING;
+
+    nm_dtls_event_do_one(ctx);
+    return NABTO_EC_OK;
+}
+
+/*
+ * Handle events for the connection phase
+ */
+void nm_dtls_event_do_one(void* data)
+{
+    struct np_dtls_cli_context* ctx = data;
+    int ret;
+    if(ctx->state == CONNECTING) {
+        ret = wolfSSL_connect(ctx->ssl);
+        if (ret != WOLFSSL_SUCCESS) {
+            int err = wolfSSL_get_error(ctx->ssl, ret);
+            if (err == WOLFSSL_ERROR_WANT_READ) {
+                NABTO_LOG_TRACE(LOG, "Want Read");
+                set_timeout(ctx);
+            } else if (err == WOLFSSL_ERROR_WANT_WRITE){
+                NABTO_LOG_TRACE(LOG, "Want Write");
+                // Wait for IO to happen
+            } else {
+                enum np_dtls_cli_event event = NP_DTLS_CLI_EVENT_CLOSED;
+                char buf[80];
+                wolfSSL_ERR_error_string(err, buf);
+                if (err > MIN_CODE_E) { // All wolfCrypt errors range ]0...MIN_CODE_E[
+                NABTO_LOG_ERROR(
+                    LOG, "Certificate verification failed: (%d) %s", err, buf);
+                    event = NP_DTLS_CLI_EVENT_CERTIFICATE_VERIFICATION_FAILED;
+                } else if( err == FATAL_ERROR ) {
+                    WOLFSSL_ALERT_HISTORY h;
+                    wolfSSL_get_alert_history(ctx->ssl, &h);
+                    if (h.last_rx.code == access_denied) {
+                        event = NP_DTLS_CLI_EVENT_ACCESS_DENIED;
+                    NABTO_LOG_ERROR(
+                    LOG, "Server returned access denied: (%d) %s" , err, buf);
+                    }
+                }
+                NABTO_LOG_INFO( LOG, "wolfssl_connect returned %d, which is %d, %s", ret , err, buf);
+                ctx->state = CLOSING;
+                np_event_queue_cancel_event(&ctx->pl->eq, ctx->timerEvent);
+                ctx->eventHandler(event, ctx->callbackData);
+                return;
+            }
+        } else if (ret == WOLFSSL_SUCCESS) {
+            char* protocol_name;
+            word16 protocol_nameSz = 0;
+            if (wolfSSL_ALPN_GetProtocol(ctx->ssl, &protocol_name, &protocol_nameSz) != SSL_SUCCESS) {
```

Comment:
is it neccessary? will it automatically fail because of wolfSSL_UseALPN(ssl, (char *)(alpnList), strlen(alpnList), WOLFSSL_ALPN_FAILED_ON_MISMATCH)?